sympy/tensor/tests/test_tensor_operators.py

```diff
@@ -15,39 +17,54 @@
 H = TensorHead("H", [L, L])
 
 
+def test_invalid_partial_derivative_valence():
+    raises(ValueError, lambda: PartialDerivative(C(j), D(-j)))
+    raises(ValueError, lambda: PartialDerivative(C(-j), D(j)))
+
+
 def test_tensor_partial_deriv():
     # Test flatten:
-    expr = PartialDerivative(PartialDerivative(A(i), A(j)), A(-i))
+    expr = PartialDerivative(PartialDerivative(A(i), A(j)), A(i))
     assert expr.expr == A(L_0)
-    assert expr.variables == (A(j), A(-L_0))
+    assert expr.variables == (A(j), A(L_0))
 
     expr1 = PartialDerivative(A(i), A(j))
     assert expr1.expr == A(i)
     assert expr1.variables == (A(j),)
 
     expr2 = A(i)*PartialDerivative(H(k, -i), A(j))
-    assert expr2.get_indices() == [L_0, k, -L_0, j]
+    assert expr2.get_indices() == [L_0, k, -L_0, -j]
+
+    expr2b = A(i)*PartialDerivative(H(k, -i), A(-j))
+    assert expr2b.get_indices() == [L_0, k, -L_0, j]
 
     expr3 = A(i)*PartialDerivative(B(k)*C(-i) + 3*H(k, -i), A(j))
-    assert expr3.get_indices() == [L_0, k, -L_0, j]
+    assert expr3.get_indices() == [L_0, k, -L_0, -j]
+
+    expr4 = (A(i) + B(i))*PartialDerivative(C(j), D(j))
+    assert expr4.get_indices() == [i, L_0, -L_0]
 
-    expr4 = (A(i) + B(i))*PartialDerivative(C(-j), D(j))
-    assert expr4.get_indices() == [i, -L_0, L_0]
+    expr4b = (A(i) + B(i))*PartialDerivative(C(-j), D(-j))
+    assert expr4b.get_indices() == [i, -L_0, L_0]
 
     expr5 = (A(i) + B(i))*PartialDerivative(C(-i), D(j))
-    assert expr5.get_indices() == [L_0, -L_0, j]
+    assert expr5.get_indices() == [L_0, -L_0, -j]
 
 
 def test_replace_arrays_partial_derivative():
     x, y, z, t = symbols("x y z t")
 
+    expr = PartialDerivative(A(i), A(-j))
+    assert expr.replace_with_arrays({A(i): [x, y], L: diag(1, 1)}, [i, j]) == Array([[1, 0], [0, 1]])
+
     expr = PartialDerivative(A(i), A(j))
-    assert expr.replace_with_arrays({A(i): [x, y]}, [i, j]) == Array([[1, 0], [0, 1]])
+    assert expr.replace_with_arrays({A(i): [x, y], L: diag(1, 1)}, [i, j]) == Array([[1, 0], [0, 1]])
+    assert expr.replace_with_arrays({A(i): [x, y], L: diag(1, -1)}, [i, j]) == Array([[1, 0], [0, -1]])
```

Comment:
@joha2 these are examples of replacing the tensorial expression with arrays (sometimes you may wish to switch to the array expression). Do you think they are correct?

Reply:
@Upabjojr for the example in line 57 I am not sure, since you derive a contravariant tensor by its covariant components, at least there should arise some metric components. In line 60 on the other hand there should arise strictly the Kronecker delta, independently of the metric. So maybe for a thorough test, one should use a non-trivial metric to distuingish these two cases.

Reply:
All other (abstract) tests are okay.

Reply:
You're right. This is probably wrong.

Reply:
> for the example in line 57 I am not sure, since you derive a contravariant tensor by its covariant components, at least there should arise some metric components.

The metric is passed with `L: diag( ... )`. The operator `diag( ... )` creates a diagonal matrix.